lib/Base/Organization/Collection.vala

```diff
@@ -16,192 +16,313 @@
    *
-   * @param a The first object to compare.
-   * @param b The second object to compare.
+   * @return How the iterators are sorted (positive when a before b, negative when b before a).
+   */
+  protected abstract int sort_func (SequenceIter<T> a, SequenceIter<T> b);
+
+  /**
+   * Run after an change to the items of the collection.
    *
-   * @return How to sort both posts (positive when a before b, negative when b before a).
+   * @param position The position of the change.
+   * @param removed The number of items that were removed.
+   * @param added The number of items that were added.
    */
-  protected int compare_items (Object a, Object b) {
-    // Sort PseudoItems according to their index
-    if (a is PseudoItem && b is PseudoItem) {
-      // Retrieve the items
-      var pseudo_a = a as PseudoItem;
-      var pseudo_b = b as PseudoItem;
+  protected virtual void after_update (uint position, uint removed, uint added) {
+    // Send the update signal
+    items_changed (position, removed, added);
+  }
 
-      // Sort the items using the set index
-      uint x = pseudo_a.index;
-      uint y = pseudo_b.index;
-      return (int) (x > y) - (int) (x < y);
+  /**
+   * Validates if the iterator cache is still valid after an change to the collection.
+   *
+   * @param position The position at which the collection was changed.
+   */
+  private void validate_cache (uint position) {
+    if (position <= last_position) {
+      last_position_valid = false;
+      last_iterator = null;
+      last_position = 0;
     }
+  }
 
-    // Sort Post and not Post
-    if (a is Post && ! (b is Post)) {
-      return 1;
-    }
-    if (! (a is Post) && b is Post) {
-      return -1;
+  /**
+   * An Iterator to enable foreach loops.
+   */
+  public class Iterator <T> : Object {
+
+    /**
+     * Constructs a new Iterator.
+     */
+    internal Iterator (Collection collection) {
+      iter = collection.items.get_begin_iter ();
+      first_next = true;
     }
 
-    // Sort Posts
-    if (a is Post && b is Post) {
-      // Retrieve the posts
-      var post_a = a as Post;
-      var post_b = b as Post;
-
-      // Retrieve the direct parents for comparison
-      var a_parent = direct_parent (post_a);
-      var b_parent = direct_parent (post_b);
-
-      // Check if one post replies to the other
-      bool a_replied_b = a_parent.replied_to_id == b_parent.id;
-      bool b_replied_a = b_parent.replied_to_id == a_parent.id;
-      if (a_replied_b || b_replied_a) {
-        return (int) (a_replied_b) - (int) (b_replied_a);
+    /**
+     * Moves to the next value.
+     *
+     * @return If a next value exists.
+     */
+    public bool next () {
+      assert (iter != null);
+      // The Vala iterator is running first next, then get.
+      // In order to also get the first item, the SequenceIter
+      // needs to wait one round before running next.
+      if (first_next) {
+        first_next = false;
```

Comment:
This feels a bit odd. But I'm not familiar with GLib iterators. The end iterator has a concept of "after the last item", so I'm surprised there isn't a similar one for "before the first item" (which is what the consumer appears to be assuming, if it's always calling `next()` as its first operation)

But I'm assuming you've looked in to it. And the "is end" should halt correctly.

Reply:
Well, having the begin iterator point to the first item allows it to be used in a for-loop, as the example in [the docs](https://valadoc.org/glib-2.0/GLib.Sequence.get_begin_iter.html) show. Seems like that was the idea why `GLib.Sequence` does it this way.